File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/SettingsFragment.java
package com.remotearthsolutions.expensetracker.fragments;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.Nullable;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.CustomEvent;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.utils.Constants;
import com.remotearthsolutions.expensetracker.utils.SharedPreferenceUtils;

public class SettingsFragment extends PreferenceFragmentCompat {

    private Context context;
    private Resources resources;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.context = context;
        resources = context.getResources();
    }

    public SettingsFragment() {
    }

    private SharedPreferences.OnSharedPreferenceChangeListener preferenceChangeListener;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.settingspreference);

        Preference preferencePeriod = findPreference(Constants.PREF_PERIOD);
        preferencePeriod.setSummary(SharedPreferenceUtils.getInstance(context).getString(Constants.PREF_PERIOD, Constants.KEY_DAILY));
        Preference preferenceCurrency = findPreference(Constants.PREF_CURRENCY);
        preferenceCurrency.setSummary(SharedPreferenceUtils.getInstance(context).getString(Constants.PREF_CURRENCY, resources.getString(R.string.default_currency)));

        preferenceChangeListener = (sharedPreferences, key) -> {

            if (key.equals(Constants.PREF_CURRENCY)) {

                Preference currencyPreference = findPreference(key);
                String val = sharedPreferences.getString(key, resources.getString(R.string.default_currency));
                currencyPreference.setSummary(val);
                Answers.getInstance().logCustom(new CustomEvent(val));

            } else if (key.equals(Constants.PREF_PERIOD)) {

                Preference periodPreference = findPreference(key);
                periodPreference.setSummary(sharedPreferences.getString(key, Constants.KEY_DAILY));

            }
        };

        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(preferenceChangeListener);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        view.setBackgroundColor(resources.getColor(android.R.color.white));

        return view;
    }

    @Override
    public void onStop() {
        super.onStop();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(preferenceChangeListener);
    }

}

File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/SettingsFragment.kt
package com.remotearthsolutions.expensetracker.fragments

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.OnBackPressedCallback
import androidx.core.content.ContextCompat
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.activities.MainActivity
import com.remotearthsolutions.expensetracker.utils.Constants
import com.remotearthsolutions.expensetracker.utils.FirebaseEventLogUtils.logCustom
import com.remotearthsolutions.expensetracker.utils.SharedPreferenceUtils
import com.remotearthsolutions.expensetracker.utils.Utils.getFlagDrawable
import kotlinx.android.synthetic.main.activity_main.*


class SettingsFragment : PreferenceFragmentCompat() {

    private var preferenceChangeListener: OnSharedPreferenceChangeListener? = null

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.settingspreference)
        val preferencePeriod =
            findPreference<Preference>(Constants.PREF_PERIOD)
        preferencePeriod!!.summary = SharedPreferenceUtils.getInstance(context!!)!!.getString(
            Constants.PREF_PERIOD,
            resources.getString(R.string.daily)
        )

        val preferenceTimeFormat = findPreference<Preference>(Constants.PREF_TIME_FORMAT)
        preferenceTimeFormat!!.summary = SharedPreferenceUtils.getInstance(context!!)!!.getString(
            Constants.PREF_TIME_FORMAT,
            resources.getString(R.string.default_time_format)
        )

        val preferenceCurrency =
            findPreference<Preference>(Constants.PREF_CURRENCY)
        val currencyName = SharedPreferenceUtils.getInstance(context!!)?.getString(
            Constants.PREF_CURRENCY,
            resources.getString(R.string.default_currency)
        )
        preferenceCurrency!!.summary = currencyName
        preferenceCurrency.setIcon(
            getFlagDrawable(
                context!!
            )
        )
        preferenceChangeListener =
            OnSharedPreferenceChangeListener { sharedPreferences: SharedPreferences, key: String ->
                when (key) {
                    Constants.PREF_CURRENCY -> {
                        val currencyPreference =
                            findPreference<Preference>(key)
                        val `val` = sharedPreferences.getString(
                            key,
                            resources.getString(R.string.default_currency)
                        )
                        currencyPreference!!.summary = `val`
                        currencyPreference.setIcon(
                            getFlagDrawable(
                                context!!
                            )
                        )
                        logCustom(context!!, `val`)
                    }
                    Constants.PREF_PERIOD -> {
                        val periodPreference =
                            findPreference<Preference>(key)
                        periodPreference!!.summary = sharedPreferences.getString(
                            key, resources.getString(R.string.daily)
                        )
                        logCustom(context!!, key)
                    }
                    Constants.PREF_TIME_FORMAT -> {
                        val timeFormatPreference =
                            findPreference<Preference>(key)
                        timeFormatPreference!!.summary =
                            sharedPreferences.getString(
                                key,
                                resources.getString(R.string.default_time_format)
                            )
                        logCustom(context!!, key)
                    }
                }
            }
        preferenceScreen.sharedPreferences
            .registerOnSharedPreferenceChangeListener(preferenceChangeListener)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        context?.let {
            view!!.setBackgroundColor(ContextCompat.getColor(context!!, android.R.color.white))
        }
        registerBackButton()
        return view
    }

    override fun onStop() {
        super.onStop()
        preferenceScreen.sharedPreferences
            .unregisterOnSharedPreferenceChangeListener(preferenceChangeListener)
    }

    fun registerBackButton(callBack: OnBackPressedCallback? = null) {
        val activity = requireActivity()
        val defaultCallback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                val fragmentManager = activity.supportFragmentManager
                val ft = fragmentManager.beginTransaction()
                ft.setCustomAnimations(R.anim.slide_in_up, 0, 0, R.anim.slide_out_down)
                ft.remove(this@SettingsFragment)
                fragmentManager.popBackStack()
                ft.commit()
                activity.toolbar!!.title = getString(R.string.title_home)
                (activity as MainActivity).hideBackButton()
            }
        }
        activity.onBackPressedDispatcher.addCallback(this, callBack ?: defaultCallback)
    }
}